### PR TITLE
Style rewrite

### PR DESCRIPTION
**Pull request by BDeveau**
Redesign of SHP UI and styling system to meet Jana's specs.

- New Format dropdown is a clone of the original Format dropdown functionality, but displays items for only the numberedElements list (can leave out H1) and additionally adjusts level on numbered items.

- New Styles dropdown to choose numbering style instead of the dialog box. Styles are currently applied globally

### CODE REVIEW

File: structuredheadings/dialogs/selectstyle.js [hunk 1]
```diff
@@ -1,150 +0,0 @@
-(function () {
```
Comment (kylegibson-rldatix) [inline structuredheadings/dialogs/selectstyle.js:1]: Why is this entire file deleted?

File: tests/structuredheadings/autonumber.js [hunk 2]
```diff
@@ -1,114 +0,0 @@
-/* bender-tags: editor,unit */
```
Comment (kylegibson-rldatix) [inline tests/structuredheadings/autonumber.js:1]: Could you explain why this file is deleted? I understand that these are only tests.

File: structuredheadings/plugin.js [hunk 3]
```diff
@@ -28,26 +38,56 @@
...
     for (var styleName in editor.config.autonumberStyles) {
       var style = editor.config.autonumberStyles[styleName];
-      for (var className in style) {
-        element.removeClass(style[className]);
+      if (typeof style === "object") {
+        for (var className in style) {
+          element.removeClass(style[className]);
+        }
+      } else {
+        element.removeClass(style);
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js:73]: What would style be in this case? Does this fail if style is null?

File: structuredheadings/plugin.js [hunk 4]
```diff
@@ -222,24 +183,162 @@
...
+              elementStyles[ editor.config.numberedElements[key] ].buildPreview(
+                "Header " + editor.config.numberedElements[key].slice(1)
+              ),
+              "Header " + editor.config.numberedElements[key].slice(1));
+          }
+          this.add("pre",
+              elementStyles.pre.buildPreview(
+                  "Formatted Text"
+                ),
+            "Preformatted Text"
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js]: The indentation here is weird

File: structuredheadings/plugin.js [hunk 5]
```diff
@@ -222,24 +183,162 @@
...
+          for (var key in editor.config.numberedElements) {
+            this.add(editor.config.numberedElements[key],
+              elementStyles[ editor.config.numberedElements[key] ].buildPreview(
+                "Header " + editor.config.numberedElements[key].slice(1)
+              ),
+              "Header " + editor.config.numberedElements[key].slice(1));
+          }
+          this.add("pre",
+              elementStyles.pre.buildPreview(
+                  "Formatted Text"
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js]: This looks like it could fit on one line

File: structuredheadings/plugin.js [hunk 6]
```diff
@@ -222,24 +183,162 @@
...
+          attributes: { "aria-label": "Numbering Formats" }
+        },
+
+        init: function () {
+
+          this.startGroup("Formats");
+
+          this.add("p",
+            elementStyles.p.buildPreview(
+              "Normal Text"
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js]: Not a big issue, but maybe put this on one line

File: structuredheadings/plugin.js [hunk 7]
```diff
@@ -222,24 +183,162 @@
...
+          editor.execCommand("reapplyStyle");
+          if (value === "p") {
+            this.setValue(value, "Normal Text");
+          } else if (value === "pre") {
+            this.setValue(value, "Formatted Text");
           } else {
-            this.setState(CKEDITOR.TRISTATE_OFF);
+            this.setValue(value, "Header " + editor.config.numberedElements[
+                editor.config.numberedElements.indexOf(value)
+              ].slice(1));
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js]: Is this indentation correct?

File: structuredheadings/plugin.js [hunk 8]
```diff
@@ -222,24 +183,162 @@
...
+            var currentTag = this.getValue();
+            var elementPath = ev.data.path;
+            var elementList = editor.config.numberedElements.concat(["pre", "p", "div"]);
+            for (var tag in elementList) {
+              if (elementStyles[elementList[tag]].checkActive(elementPath, editor)) {
+                if (elementList[tag] !== currentTag) {
+                  if (elementList[tag] === "p") {
+                    this.setValue(elementList[tag], "Normal Text");
+                  } else if (elementList[tag] === "pre") {
+                    this.setValue(elementList[tag], "Formatted Text");
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js:247]: Maybe we could have an object mapping for these labels?

File: structuredheadings/plugin.js [hunk 9]
```diff
@@ -222,24 +183,162 @@
...
+                  } else {
+                    this.setValue(elementList[tag], "Header " +
+                          editor.config.numberedElements[tag].slice(1));
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js]: maybe this?

```javascript
this.setValue(
    elementList[tag],
    "Header " + editor.config.numberedElements[tag].slice(1)
);
```

File: structuredheadings/plugin.js [hunk 10]
```diff
@@ -222,24 +183,162 @@
...
+                  }
+                }
+                return;
+              }
+            }
+
+            this.setValue("");
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js:259]: What does this accomplish?

File: structuredheadings/plugin.js [hunk 11]
```diff
@@ -222,24 +183,162 @@
...
+        },
+
+        onClick: function (value) {
+          if (value === "restart") {
+            editor.execCommand("restartNumbering");
+          } else if (value === "clear") {
+            clearStyles(editor, editor.elementPath().block);
+            clearLevel(editor, editor.elementPath().block);
+            clearNumbering(editor, editor.elementPath().block);
+          } else {
```
Comment (kylegibson-rldatix) [inline structuredheadings/plugin.js:297]: What's the else value?

File: structuredheadings/dialogs/selectstyle.js (hunk 1 above)
Comment (BDeveau) [reply structuredheadings/dialogs/selectstyle.js:1]: The plugin no longer has a dialog

File: tests/structuredheadings/autonumber.js (hunk 2 above)
Comment (BDeveau) [reply tests/structuredheadings/autonumber.js:1]: Tests involve a command that is no longer used. SHP underwent a major rewrite

**Comment by kylegibson-rldatix** [comment]
I feel that I'm missing a lot of context for the changes here. I did the best I could with what I understood. I don't fully understand what the purpose of this code is, or the context, so please help me understand.

File: structuredheadings/plugin.js (hunk 3 above)
Comment (BDeveau) [reply structuredheadings/plugin.js:73]: The only way for style to be null is if someone edited the config and supplied blank styles to overwrite the default values. If that were the case, style being null would be the least of your problems, the editor likely wouldn't load.

In this case, style is a CSS class

File: structuredheadings/plugin.js (hunk 6 above)
Comment (BDeveau) [reply structuredheadings/plugin.js]: done

File: structuredheadings/plugin.js (hunk 5 above)
Comment (BDeveau) [reply structuredheadings/plugin.js]: done

File: structuredheadings/plugin.js (hunk 7 above)
Comment (BDeveau) [reply structuredheadings/plugin.js]: updated, expanded... made sure things are aligned, I'm not sure what correct indentation is, I have the list index split, the closing bracket aligned with the start of the line where it is opened.

File: structuredheadings/plugin.js (hunk 8 above)
Comment (BDeveau) [reply structuredheadings/plugin.js:247]: elementList is referencing config.numberedElements which needs to be a flat array for purposes of integer indexes. In this case these tags are being added with .concat within this function and aren't a part of the original list. Setting up an object would be taking place still within this function a couple lines up and I don't think would be a benefit.

File: structuredheadings/plugin.js (hunk 9 above)
Comment (BDeveau) [reply structuredheadings/plugin.js]: done

File: structuredheadings/plugin.js (hunk 10 above)
Comment (BDeveau) [reply structuredheadings/plugin.js:259]: If the current selection is styled, the name of the dropdown box is updated to the active element style. If the selection doesn't match a known style, this line defaults the select box to no selection and defaults the dropdown name.

File: structuredheadings/plugin.js (hunk 11 above)
Comment (BDeveau) [reply structuredheadings/plugin.js:297]: restart and clear are hardcoded entries to the select element with specific commands, the else accounts for each of the stylenames which are dynamic, based on config setting and applies that style.

**Comment by BDeveau** [comment]
@kylegibson At the end of last week this plugin underwent a major rewrite, rendering a lot of previous code obsolete and unused. Changes here are extensive, eliminating editor commands, dialog boxes, and combining what used to be multiple actions into one, all bundled in a UI change from buttons to a select dropdown.

The main context with the most recent change was 4 or 5 toolbar buttons with individual funcionaltiy are being mashed into one singular select dropdown and one onSelect event.

File: structuredheadings/dialogs/selectstyle.js (hunk 1 above)
Comment (kylegibson-rldatix) [reply structuredheadings/dialogs/selectstyle.js:1]: Ah, that makes sense

**Review by kylegibson-rldatix — APPROVED** [review]
Looks good

**Pull request merged**